### PR TITLE
[INTERNAL] Refactor requires build/requires dependencies determination

### PR DESCRIPTION
**Pull request by RandomByte**
* Move requiresBuild() logic from TaskRunner to ProjectBuildContext
    * This likely needs to be moved into a module of its own at some
      point. For now, TaskRunner was the wrong place. If a project does
      not need to be built, there is no need to instantiate a TaskRunner
      in the first place. This is now implemented.
* TaskRunner can now compile a list of all direct project dependencies
  required by the set of tasks that it will execute for the given build
  options.
* Custom tasks defining specVersion >= 3.0 can now provide a
  'determineRequiredDependencies' callback to define which dependencies
  they require access to.

JIRA: CPOUI5FOUNDATION-568

### CODE REVIEW

File: lib/build/TaskRunner.js [hunk 1]
```diff
@@ -120,17 +122,19 @@ class TaskRunner {
...
 		}
 	}
 
 	/**
-	 * Takes a list of tasks which should be executed from the available task list of the current builder
+	 * First compiles a list of all tasks that will be executed, then a list of all direct project
+	 * dependencies that those tasks require access to.
 	 *
-	 * @returns {Promise} Returns promise chain with tasks
+	 * @returns {Set} Returns a set containing the names of all required direct project dependencies
```
Comment (matz3) [inline lib/build/TaskRunner.js:134]:
```suggestion
	 * @returns {Set<string>} Returns a set containing the names of all required direct project dependencies
```

**Review by matz3 — COMMENTED** [review]
Nice 👍🏻

File: test/lib/build/TaskRunner.js [hunk 2]
```diff
@@ -2,6 +2,7 @@ import test from "ava";
 import sinonGlobal from "sinon";
 import esmock from "esmock";
 import logger from "@ui5/logger";
+logger.setLevel("perf");
```
Comment (flovogt) [inline test/lib/build/TaskRunner.js:5]: why do we have to set the log level here?

File: lib/build/TaskRunner.js [hunk 3]
```diff
@@ -214,138 +200,211 @@ class TaskRunner {
...
+		this._tasks[newTaskName] = {
+			task: this._createCustomTaskWrapper({
+				task,
+				project,
+				taskUtil,
+				taskName: newTaskName,
+				taskConfiguration: taskDef.configuration,
+				getDependenciesReader: () => {
+					// Create the dependencies reader on-demand
+					return this._createDependencyReader(requiredDependencies);
```
[the remark this answers is not in the document]
Comment (RandomByte) [reply lib/build/TaskRunner.js]: Done

File: test/lib/build/TaskRunner.js (hunk 2 above)
Comment (RandomByte) [reply test/lib/build/TaskRunner.js:5]: In function `_executeTask`, we have some code that is only executed when log-level "perf" is activated

**Review by flovogt — APPROVED** [review]

**Review by matz3 — APPROVED** [review]

**Pull request merged**